core/common/src/main/java/alluxio/AbstractClient.java

```diff
@@ -196,14 +196,11 @@ public synchronized void connect() throws IOException, ConnectionFailedException
   public synchronized void disconnect() {
     if (mConnected) {
       LOG.debug("Disconnecting from the {} {} {}", getServiceName(), mMode, mAddress);
-      mConnected = false;
-    }
-    try {
       beforeDisconnect();
       if (mProtocol != null) {
```

Comment:
as per you comment in my original PR, I don't think it can ever be the case that `mProtocol` is null and `mConnected == true`, so I would not do this check here because it can be masking an invariant violation